File: taurus-protect-sdk-python/taurus_protect/services/taurus_network/participant_service.py
"""Participant service for Taurus Network in Taurus-PROTECT SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, List, Optional

from taurus_protect.mappers.taurus_network.participant import (
    my_participant_from_dto,
    participant_from_dto,
    participants_from_dto,
)
from taurus_protect.models.taurus_network.participant import (
    CreateParticipantAttributeRequest,
    GetParticipantOptions,
    ListParticipantsOptions,
    MyParticipant,
    Participant,
)
from taurus_protect.services._base import BaseService

if TYPE_CHECKING:
    from taurus_protect._internal.openapi.exceptions import ApiException


class ParticipantService(BaseService):
    """
    Service for Taurus Network participant operations.

    Provides methods to retrieve and manage Taurus Network participants,
    including the current tenant's participant information and attributes.

    Example:
        >>> # Get my participant info
        >>> my_participant = client.taurus_network.participants.get_my_participant()
        >>> print(f"My participant: {my_participant.participant.name}")
        >>>
        >>> # List all visible participants
        >>> participants = client.taurus_network.participants.list()
        >>> for p in participants:
        ...     print(f"{p.name}: {p.country}")
        >>>
        >>> # Get specific participant
        >>> opts = GetParticipantOptions(include_total_pledges_valuation=True)
        >>> participant = client.taurus_network.participants.get("participant-id", opts)
    """

    def __init__(
        self,
        api_client: Any,
        participant_api: Any,
    ) -> None:
        """
        Initialize participant service.

        Args:
            api_client: The OpenAPI client instance.
            participant_api: The TaurusNetworkParticipantApi service from OpenAPI client.
        """
        super().__init__(api_client)
        self._participant_api = participant_api

    def get_my_participant(self) -> MyParticipant:
        """
        Get the current participant with settings.

        Returns the participant linked to the current tenant along with
        their Taurus Network settings.

        Returns:
            MyParticipant containing the participant and settings.

        Raises:
            NotFoundError: If the current tenant is not a participant.
            APIError: If API request fails.
        """
        try:
            resp = self._participant_api.taurus_network_service_get_my_participant()

            result = my_participant_from_dto(resp)
            if result is None:
                from taurus_protect.errors import NotFoundError

                raise NotFoundError("My participant not found")

            return result
        except Exception as e:
            from taurus_protect._internal.openapi.exceptions import ApiException
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            if isinstance(e, ApiException):
                raise self._handle_error(e) from e
            raise self._handle_error(e) from e

    def get(
        self,
        participant_id: str,
        options: Optional[GetParticipantOptions] = None,
    ) -> Participant:
        """
        Get a participant by ID.

        Args:
            participant_id: The participant ID to retrieve.
            options: Optional retrieval options.

        Returns:
            The participant.

        Raises:
            ValueError: If participant_id is empty.
            NotFoundError: If participant not found.
            APIError: If API request fails.
        """
        self._validate_required(participant_id, "participant_id")
        opts = options or GetParticipantOptions()

        try:
            resp = self._participant_api.taurus_network_service_get_participant(
                participant_id=participant_id,
                include_total_pledges_valuation=(
                    opts.include_total_pledges_valuation
                    if opts.include_total_pledges_valuation
                    else None
                ),
            )

            result = getattr(resp, "result", None)
            if result is None:
                from taurus_protect.errors import NotFoundError

                raise NotFoundError(f"Participant {participant_id} not found")

            participant = participant_from_dto(result)
            if participant is None:
                from taurus_protect.errors import NotFoundError

                raise NotFoundError(f"Participant {participant_id} not found")

            return participant
        except Exception as e:
            from taurus_protect._internal.openapi.exceptions import ApiException
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            if isinstance(e, ApiException):
                raise self._handle_error(e) from e
            raise self._handle_error(e) from e

    def list(
        self,
        options: Optional[ListParticipantsOptions] = None,
    ) -> List[Participant]:
        """
        List visible Taurus Network participants.

        Returns participants that are visible to the current tenant's
        participant based on allowed interactions.

        Args:
            options: Optional filtering options.

        Returns:
            List of visible participants.

        Raises:
            APIError: If API request fails.
        """
        opts = options or ListParticipantsOptions()

        try:
            resp = self._participant_api.taurus_network_service_get_participants(
                participant_ids=opts.participant_ids,
                include_total_pledges_valuation=(
                    opts.include_total_pledges_valuation
                    if opts.include_total_pledges_valuation
                    else None
                ),
            )

            result = getattr(resp, "result", None)
            return participants_from_dto(result) if result else []
        except Exception as e:
            from taurus_protect._internal.openapi.exceptions import ApiException
            from taurus_protect.errors import APIError

            if isinstance(e, APIError):
                raise
            if isinstance(e, ApiException):
                raise self._handle_error(e) from e
            raise self._handle_error(e) from e

    def create_participant_attribute(
        self,
        participant_id: str,
        request: CreateParticipantAttributeRequest,
    ) -> None:
        """
        Create an attribute for a participant.

        Creates a new attribute for the specified participant. The attribute
        can optionally be shared to the linked Taurus Network participant.

        Args:
            participant_id: The participant ID.
            request: The attribute creation request.

        Raises:
            ValueError: If participant_id or request is invalid.
            NotFoundError: If participant not found.
            APIError: If API request fails.

        Note:
            Required role: Admin.
        """
        self._validate_required(participant_id, "participant_id")
        if request is None:
            raise ValueError("request cannot be None")
        self._validate_required(request.key, "key")
        self._validate_required(request.value, "value")

        try:
            # Import the body model
            from taurus_protect._internal.openapi.models.taurus_network_service_create_participant_attribute_body import (
                TaurusNetworkServiceCreateParticipantAttributeBody,
            )
            from taurus_protect._internal.openapi.models.tgvalidatord_participant_attribute_data import (
                TgvalidatordParticipantAttributeData,
            )

            # Build the attribute data
            attribute_data = TgvalidatordParticipantAttributeData(
                key=request.key,
                value=request.value,
                content_type=request.content_type if request.content_type else None,
                type=request.attribute_type if request.attribute_type else None,
                subtype=request.subtype if request.subtype else None,
            )

            # Build the request body
            body = TaurusNetworkServiceCreateParticipantAttributeBody(
                attribute_data=attribute_data,
                share_to_taurus_network_participant=request.share_to_taurus_network_participant,
            )

            self._participant_api.taurus_network_service_create_participant_attribute(
                participant_id=participant_id,
                body=body,
            )
        except Exception as e:
            from taurus_protect._internal.openapi.exceptions import ApiException
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            if isinstance(e, ApiException):
                raise self._handle_error(e) from e
            raise self._handle_error(e) from e

    def delete_participant_attribute(
        self,
        participant_id: str,
        attribute_id: str,
    ) -> None:
        """
        Delete an attribute for a participant.

        Deletes the specified attribute from the participant. If the attribute
        was shared to the linked Taurus Network participant, a notification
        will be sent.

        Args:
            participant_id: The participant ID.
            attribute_id: The attribute ID to delete.

        Raises:
            ValueError: If participant_id or attribute_id is empty.
            NotFoundError: If participant or attribute not found.
            APIError: If API request fails.

        Note:
            Required role: Admin.
        """
        self._validate_required(participant_id, "participant_id")
        self._validate_required(attribute_id, "attribute_id")

        try:
            self._participant_api.taurus_network_service_delete_participant_attribute(
                participant_id=participant_id,
                attribute_id=attribute_id,
            )
        except Exception as e:
            from taurus_protect._internal.openapi.exceptions import ApiException
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            if isinstance(e, ApiException):
                raise self._handle_error(e) from e
            raise self._handle_error(e) from e

File: taurus-protect-sdk-typescript/src/services/taurus-network/participant-service.ts
/**
 * Participant service for Taurus Network in Taurus-PROTECT SDK.
 *
 * Provides methods for managing Taurus Network participants,
 * including the current tenant's participant information and attributes.
 */

import { NotFoundError, ValidationError } from '../../errors';
import type { TaurusNetworkParticipantApi } from '../../internal/openapi/apis/TaurusNetworkParticipantApi';
import type {
  TgvalidatordTnParticipant,
  TgvalidatordTnParticipantSettings,
} from '../../internal/openapi/models/index';
import { BaseService } from '../base';

/**
 * Participant information from Taurus Network.
 */
export interface Participant {
  id: string;
  name: string;
  legalAddress?: string;
  country?: string;
  logoBase64?: string;
  publicKey?: string;
  shield?: string;
  publicSubname?: string;
  legalEntityIdentifier?: string;
  status?: string;
  originRegistrationDate?: Date;
  originDeletionDate?: Date;
  createdAt?: Date;
  updatedAt?: Date;
  ownedSharedAddressesCount?: string;
  targetedSharedAddressesCount?: string;
  outgoingTotalPledgesValuationBaseCurrency?: string;
  incomingTotalPledgesValuationBaseCurrency?: string;
  attributes?: ParticipantAttribute[];
}

/**
 * Participant attribute.
 */
export interface ParticipantAttribute {
  id?: string;
  key: string;
  value: string;
  contentType?: string;
  type?: string;
  subtype?: string;
  owner?: string;
  isTaurusNetworkShared?: boolean;
}

/**
 * Allowed participant for interactions.
 */
export interface AllowedParticipant {
  id?: string;
  name?: string;
}

/**
 * Participant settings for the current tenant.
 */
export interface ParticipantSettings {
  interactingAllowedCountries?: string[];
  status?: string;
  interactingAllowedParticipants?: AllowedParticipant[];
  termsAndConditionsAcceptedAt?: Date;
}

/**
 * My participant with settings.
 */
export interface MyParticipant {
  participant: Participant;
  settings?: ParticipantSettings;
}

/**
 * Options for getting a participant.
 */
export interface GetParticipantOptions {
  includeTotalPledgesValuation?: boolean;
}

/**
 * Options for listing participants.
 */
export interface ListParticipantsOptions {
  participantIds?: string[];
  includeTotalPledgesValuation?: boolean;
}

/**
 * Request to create a participant attribute.
 */
export interface CreateParticipantAttributeRequest {
  key: string;
  value: string;
  contentType?: string;
  attributeType?: string;
  subtype?: string;
  shareToTaurusNetworkParticipant?: boolean;
}

/**
 * Maps a DTO to a Participant.
 */
function participantFromDto(dto?: TgvalidatordTnParticipant): Participant | undefined {
  if (!dto) {
    return undefined;
  }

  return {
    id: dto.id ?? '',
    name: dto.name ?? '',
    legalAddress: dto.legalAddress,
    country: dto.country,
    logoBase64: dto.logoBase64,
    publicKey: dto.publicKey,
    shield: dto.shield,
    publicSubname: dto.publicSubname,
    legalEntityIdentifier: dto.legalEntityIdentifier,
    status: dto.status,
    originRegistrationDate: dto.originRegistrationDate,
    originDeletionDate: dto.originDeletionDate,
    createdAt: dto.createdAt,
    updatedAt: dto.updatedAt,
    ownedSharedAddressesCount: dto.ownedSharedAddressesCount,
    targetedSharedAddressesCount: dto.targetedSharedAddressesCount,
    outgoingTotalPledgesValuationBaseCurrency: dto.outgoingTotalPledgesValuationBaseCurrency,
    incomingTotalPledgesValuationBaseCurrency: dto.incomingTotalPledgesValuationBaseCurrency,
    attributes: dto.attributes?.map((attr) => ({
      id: attr.id,
      key: attr.key ?? '',
      value: attr.value ?? '',
      contentType: attr.contentType,
      type: attr.type,
      subtype: attr.subtype,
      owner: attr.owner,
      isTaurusNetworkShared: attr.isTaurusNetworkShared,
    })),
  };
}

/**
 * Maps settings DTO to ParticipantSettings.
 */
function settingsFromDto(dto?: TgvalidatordTnParticipantSettings): ParticipantSettings | undefined {
  if (!dto) {
    return undefined;
  }

  return {
    interactingAllowedCountries: dto.interactingAllowedCountries,
    status: dto.status,
    interactingAllowedParticipants: dto.interactingAllowedParticipants?.map((p) => ({
      id: p.id,
      name: p.name,
    })),
    termsAndConditionsAcceptedAt: dto.termsAndConditionsAcceptedAt,
  };
}

/**
 * Service for Taurus Network participant operations.
 *
 * Provides methods to retrieve and manage Taurus Network participants,
 * including the current tenant's participant information and attributes.
 *
 * @example
 * ```typescript
 * // Get my participant info
 * const myParticipant = await participantService.getMyParticipant();
 * console.log(`My participant: ${myParticipant.participant.name}`);
 *
 * // List all visible participants
 * const participants = await participantService.list();
 * for (const p of participants) {
 *   console.log(`${p.name}: ${p.country}`);
 * }
 *
 * // Get specific participant
 * const participant = await participantService.get('participant-id', {
 *   includeTotalPledgesValuation: true,
 * });
 * ```
 */
export class ParticipantService extends BaseService {
  private readonly participantApi: TaurusNetworkParticipantApi;

  /**
   * Creates a new ParticipantService instance.
   *
   * @param participantApi - The TaurusNetworkParticipantApi instance from the OpenAPI client
   */
  constructor(participantApi: TaurusNetworkParticipantApi) {
    super();
    this.participantApi = participantApi;
  }

  /**
   * Gets the current participant with settings.
   *
   * Returns the participant linked to the current tenant along with
   * their Taurus Network settings.
   *
   * @returns MyParticipant containing the participant and settings
   * @throws {@link NotFoundError} If the current tenant is not a participant
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const myParticipant = await participantService.getMyParticipant();
   * console.log(`My participant: ${myParticipant.participant.name}`);
   * console.log(`Settings: ${JSON.stringify(myParticipant.settings)}`);
   * ```
   */
  async getMyParticipant(): Promise<MyParticipant> {
    return this.execute(async () => {
      const response = await this.participantApi.taurusNetworkServiceGetMyParticipant();

      const participant = participantFromDto(response.result);
      if (!participant) {
        throw new NotFoundError('My participant not found');
      }

      return {
        participant,
        settings: settingsFromDto(response.settings),
      };
    });
  }

  /**
   * Gets a participant by ID.
   *
   * @param participantId - The participant ID to retrieve
   * @param options - Optional retrieval options
   * @returns The participant
   * @throws {@link ValidationError} If participantId is empty
   * @throws {@link NotFoundError} If participant not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const participant = await participantService.get('participant-123', {
   *   includeTotalPledgesValuation: true,
   * });
   * console.log(`Participant: ${participant.name}`);
   * ```
   */
  async get(participantId: string, options?: GetParticipantOptions): Promise<Participant> {
    if (!participantId || participantId.trim() === '') {
      throw new ValidationError('participantId is required');
    }

    return this.execute(async () => {
      const response = await this.participantApi.taurusNetworkServiceGetParticipant({
        participantID: participantId,
        includeTotalPledgesValuation: options?.includeTotalPledgesValuation,
      });

      const participant = participantFromDto(response.result);
      if (!participant) {
        throw new NotFoundError(`Participant ${participantId} not found`);
      }

      return participant;
    });
  }

  /**
   * Lists visible Taurus Network participants.
   *
   * Returns participants that are visible to the current tenant's
   * participant based on allowed interactions.
   *
   * @param options - Optional filtering options
   * @returns List of visible participants
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * // List all participants
   * const participants = await participantService.list();
   *
   * // List specific participants with valuation
   * const filtered = await participantService.list({
   *   participantIds: ['id1', 'id2'],
   *   includeTotalPledgesValuation: true,
   * });
   * ```
   */
  async list(options?: ListParticipantsOptions): Promise<Participant[]> {
    return this.execute(async () => {
      const response = await this.participantApi.taurusNetworkServiceGetParticipants({
        participantIDs: options?.participantIds,
        includeTotalPledgesValuation: options?.includeTotalPledgesValuation,
      });

      const participants: Participant[] = [];
      if (response.result) {
        for (const dto of response.result) {
          const participant = participantFromDto(dto);
          if (participant) {
            participants.push(participant);
          }
        }
      }

      return participants;
    });
  }

  /**
   * Creates an attribute for a participant.
   *
   * Creates a new attribute for the specified participant. The attribute
   * can optionally be shared to the linked Taurus Network participant.
   *
   * @param participantId - The participant ID
   * @param request - The attribute creation request
   * @throws {@link ValidationError} If participantId or request is invalid
   * @throws {@link NotFoundError} If participant not found
   * @throws {@link APIError} If API request fails
   *
   * Note: Required role is Admin.
   *
   * @example
   * ```typescript
   * await participantService.createParticipantAttribute('participant-123', {
   *   key: 'department',
   *   value: 'treasury',
   *   shareToTaurusNetworkParticipant: true,
   * });
   * ```
   */
  async createParticipantAttribute(
    participantId: string,
    request: CreateParticipantAttributeRequest
  ): Promise<void> {
    if (!participantId || participantId.trim() === '') {
      throw new ValidationError('participantId is required');
    }
    if (!request.key || request.key.trim() === '') {
      throw new ValidationError('key is required');
    }
    if (!request.value || request.value.trim() === '') {
      throw new ValidationError('value is required');
    }

    return this.execute(async () => {
      await this.participantApi.taurusNetworkServiceCreateParticipantAttribute({
        participantID: participantId,
        body: {
          attributeData: {
            key: request.key,
            value: request.value,
            contentType: request.contentType,
            type: request.attributeType,
            subtype: request.subtype,
          },
          shareToTaurusNetworkParticipant: request.shareToTaurusNetworkParticipant,
        },
      });
    });
  }

  /**
   * Deletes an attribute for a participant.
   *
   * Deletes the specified attribute from the participant. If the attribute
   * was shared to the linked Taurus Network participant, a notification
   * will be sent.
   *
   * @param participantId - The participant ID
   * @param attributeId - The attribute ID to delete
   * @throws {@link ValidationError} If participantId or attributeId is empty
   * @throws {@link NotFoundError} If participant or attribute not found
   * @throws {@link APIError} If API request fails
   *
   * Note: Required role is Admin.
   *
   * @example
   * ```typescript
   * await participantService.deleteParticipantAttribute('participant-123', 'attr-456');
   * ```
   */
  async deleteParticipantAttribute(participantId: string, attributeId: string): Promise<void> {
    if (!participantId || participantId.trim() === '') {
      throw new ValidationError('participantId is required');
    }
    if (!attributeId || attributeId.trim() === '') {
      throw new ValidationError('attributeId is required');
    }

    return this.execute(async () => {
      await this.participantApi.taurusNetworkServiceDeleteParticipantAttribute({
        participantID: participantId,
        attributeID: attributeId,
      });
    });
  }
}
